### PR TITLE
{engine, plugin}: Propagate more options to MLCs

### PR DESCRIPTION
**Pull request by abhinav**
*Each commit may be reviewed separately*

Adds the following new options for MLC construction:

- additionalSecretOutputs
- customTimeouts
- deleteBeforeReplace
- deletedWith
- ignoreChanges
- replaceOnChanges
- retainOnDelete

The flow of this information is roughly as follows:
- ResourceMonitor gets a RegisterResource request from a Pulumi program.
  If this request is for a remote resource (an MLC),
  it converts this request into a call to a plugin.Provider object.
- The plugin.Provider receives these in a plugin.ConstructOptions
  which it converts into a ConstructRequest RPC call
  that it sends onwards to the process implementing the MLC.

Following this change, the MLC implementations in each language will
pull information from the ConstructRequest, and set options from it
appropriately, on the target resource

Note that in ResourceMonito.RegisterResource, there's a bit that goes:

    if remote {
        provider.Construct(...)
    } else {
        resource.NewGoal(...)
    }

With this change, all information passed to resource.NewGoal
is also touched for provider.Construct in some form
so nearly all options are accounted for.

Refs #12154

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12681** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12681" target="_blank">Graphite</a> 
    * **PR #12682** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12682" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/12682?utm_source=stack-comment).

**Comment by abhinav** [comment]
Open questions resolved after discussion with @justinvp: we'll use strings for CustomTimeouts fields.

Ready for review.

**Review by justinvp — APPROVED** [review]

File: pkg/resource/deploy/source_eval.go [hunk 1]
```diff
@@ -1239,11 +1227,41 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 			outputDeps[string(k)] = &pulumirpc.RegisterResourceResponse_PropertyDependencies{Urns: urns}
 		}
 	} else {
+		additionalSecretKeys := make([]resource.PropertyKey, 0, len(additionalSecretOutputs))
+		for _, name := range req.GetAdditionalSecretOutputs() {
```
[the remark this answers is not in the document]
Comment (abhinav) [reply pkg/resource/deploy/source_eval.go:1231]: No reason. Switching.

**Comment by abhinav** [comment]
bors merge

**Pull request merged**